xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java

```diff
@@ -170,31 +173,198 @@ public void onError(Status error) {
   public DownstreamTlsContext getDownstreamTlsContext(Channel channel) {
     if (curListener != null && channel != null) {
       SocketAddress localAddress = channel.localAddress();
-      checkState(
-          localAddress instanceof InetSocketAddress,
-          "Channel localAddress is expected to be InetSocketAddress");
-      InetSocketAddress localInetAddr = (InetSocketAddress) localAddress;
-      checkState(
-          port == localInetAddr.getPort(),
-          "Channel localAddress port does not match requested listener port");
-      return getDownstreamTlsContext(localInetAddr);
+      SocketAddress remoteAddress = channel.remoteAddress();
+      if (localAddress instanceof InetSocketAddress && remoteAddress instanceof InetSocketAddress) {
+        InetSocketAddress localInetAddr = (InetSocketAddress) localAddress;
+        InetSocketAddress remoteInetAddr = (InetSocketAddress) remoteAddress;
+        if (port == localInetAddr.getPort()) {
+          return getDownstreamTlsContext(localInetAddr, remoteInetAddr);
+        }
+      }
```

Comment:
Is this changed behavior documented anywhere?

Reply:
There are multiple parts to the answer:

- the `checkState` check for InetSocketAddress was not a good idea since throwing the IllegalStateException doesn't achieve anything and in both production and test code it will always be an InetSocketAddress. There probably was some code in the past where netty's `EmbeddedSocketAddress`  was used but not any more. So this is just a check before casting SocketAddress to InetSocketAddress.
- `checkState` for verifying the port is interesting: in reality we should never receive a channel where the destination port doesn't match the port my server is bound to but just in case it happens (because of a bug in netty?) should we throw an exception or return null for the tlsContext? I can see an argument for throwing an exception just to be safe so we don't accidentally fall back to plaintext. Let me think

Reply:
So this was a good point for port check and I reverted that part.